include/bout/mesh.hxx

```diff
@@ -444,11 +444,26 @@ class Mesh {
       // No coordinate system set. Create default
       // Note that this can't be allocated here due to incomplete type
       // (circular dependency between Mesh and Coordinates)
-      coords_map.emplace(location, createDefaultCoordinates(location));
+      coords_map[location] = std::shared_ptr<Coordinates>(nullptr);
+      coords_map[location] = createDefaultCoordinates(location);
       return coords_map[location].get();
     }
   }
 
+  /// Re-calculate staggered Coordinates, useful if CELL_CENTRE Coordinates are changed
+  void recalculateStaggeredCoordinates() {
```

Comment:
Another issue about checking that we don't do something unexpected: I think we expect `geometry()` to be called for staggered `Coordinates` only when they are being interpolated from the `CELL_CENTRE` `Coordinates`. I think we should do like 3. above and throw an exception if `geometry()` is called for `location != CELL_CENTRE`, unless an argument (different from the one for the exception above?) is passed to suppress the exception. We can pass that argument in the `Coordinates(Mesh*,CELL_LOC,Coordinates*)` constructor to allow it to call `geometry()`, and a user can if they *really* want to call `geometry()` on a staggered `Coordinates`.